store/store.go

```diff
@@ -550,33 +550,61 @@ func (s *Store[H]) get(ctx context.Context, hash header.Hash) ([]byte, error) {
 	return data, nil
 }
 
-// advanceContiguousHead updates contiguousHead and heightSub if a higher
-// contiguous header exists on a disk.
-func (s *Store[H]) advanceContiguousHead(ctx context.Context, height uint64) {
-	newHead := s.nextContiguousHead(ctx, height)
-	if newHead.IsZero() || newHead.Height() <= height {
-		return
+// advanceHeadAndTail moves contiguous Head and Tail if a new or older exists respectively.
+// It looks throw caches, pending headers and datastore to find the new Head and Tail.
+// TODO(@Wondertan): Beware of the performance penalty of this approach, which always makes a at least one
+// datastore lookup for both Tail and Head.
+func (s *Store[H]) advanceHeadAndTail(ctx context.Context) {
+	newHead, changed := s.nextHead(ctx)
+	if changed {
+		s.contiguousHead.Store(&newHead)
+		s.heightSub.SetHeight(newHead.Height())
+		log.Infow("new head", "height", newHead.Height(), "hash", newHead.Hash())
+		s.metrics.newHead(newHead.Height())
+	}
+
+	newTail, changed := s.nextTail(ctx)
+	if changed {
+		s.tailHeader.Store(&newTail)
+		log.Infow("new tail", "height", newTail.Height(), "hash", newTail.Hash())
+		s.metrics.newTail(newTail.Height())
+	}
+}
+
+// nextHead finds the new contiguous Head by iterating the current Head up until the newer height Head is found.
+// Returns true if the newer one was found.
+func (s *Store[H]) nextHead(ctx context.Context) (head H, changed bool) {
+	head, err := s.Head(ctx)
+	if err != nil {
```

Comment:
should we debug log some error here? In both err case for head load and tail load?

Reply:
Its unrealistic to get any error there currently, but as i am future proving by calling Head/Tail instead of sccessing pointers directly, it also makes sense to add the log there

Reply:
[8d6d875](https://github.com/celestiaorg/go-header/pull/283/commits/8d6d875c3f3ac193bf7fee4c7f1d5b311549502d)